slyd/slyd/orm/base.py

```diff
@@ -0,0 +1,550 @@
+        basic_attrs = {}
+        file_schema_attrs = {}
+
+        for attrname, value in iteritems(attrs):
+            if isinstance(value, BaseRelationship):
+                fields[attrname] = value
+            elif isinstance(value, Field):
+                if value.primary_key:
+                    if primary_key:
+                        raise ImproperlyConfigured(
+                            u"Model '{}' declared with more than one primary "
+                            u"key".format(name))
+                    primary_key = attrname
+                fields[attrname] = value
+            # move decorated marshmallow methods to the file schema
+            elif hasattr(value, '__marshmallow_tags__'):
+                file_schema_attrs[attrname] = value
+            else:
+                basic_attrs[attrname] = value
+
+        if fields and not primary_key:
+            raise ImproperlyConfigured(
+                u"Model '{}' declared with no primary key".format(name))
+
+        dependencies = {}
+        for attrname, field in iteritems(fields):
+            if attrname == primary_key:
+                dependencies[attrname] = set()
+            elif isinstance(field, DependantField):
+                dependencies[attrname] = {primary_key, field.when}
+            else:
+                dependencies[attrname] = {primary_key}
+        try:
+            ordered_fields = toposort_flatten(dependencies)
+        except ValueError as e:
+            raise ImproperlyConfigured(e.message)
+
+        cls = super(ModelMeta, mcs).__new__(mcs, name, bases, basic_attrs)
+        cls._pk_field = primary_key
+        cls._fields = fields
+        cls._ordered_fields = ordered_fields
+        cls._file_fields = file_fields = {k for k, f in iteritems(fields)
+                                          if not f.ignore_in_file}
+        cls._field_names = sorted(k for k, f in iteritems(fields)
+                                  if isinstance(f, Field))
+        cls._relationship_names = sorted(k for k, f in iteritems(fields)
+                                         if isinstance(f, BaseRelationship))
+        cls.opts = ModelOpts(meta, cls)
+        cls.collection = type(name + 'Collection', (ModelCollection,), {
+            'model': cls
+        })
+
+        for attrname, field in iteritems(fields):
+            if attrname in file_fields:
+                file_schema_attrs[attrname] = field
+            field.contribute_to_class(cls, attrname)
+
+        # build a marshmallow schema for the filesystem format
+        meta_bases = (meta, object) if meta else (object,)
+        file_schema_attrs['Meta'] = type('Meta', meta_bases, {
+            'model': cls
+        })
+        cls.file_schema = type(cls.__name__ + 'Schema', (BaseFileSchema,),
+                               file_schema_attrs)
+
+        # add new model to registry by name
+        models[name] = cls
+        return cls
+
+    @property
+    def _file_model(cls):
```

Comment:
python's property doesn't work on the class object, getting the attribute returns the property descriptor. but i've refactored this adding a custom class_property descriptor that allows the property to be read both from the class and instance, so it's better now.